### PR TITLE
Support recursive TypedDicts

### PR DESCRIPTION
**Pull request by ilevkivskyi**
This is a continuation of #13297
Depends on #13371 

It was actually quite easy, essentially just a 1-to-1 mapping from the other PR. I will rebase when the first PR will be merged.

cc @JukkaL

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
The `mypy_primer` effect is all from base PR, so the effect of this one is zero.

File: test-data/unit/check-recursive-types.test [hunk 1]
```diff
@@ -438,3 +438,230 @@ E = List[E[E[T]]]  # E: Invalid recursive alias: type variable nesting on right
...
+class TDB(TypedDict):
+    xb: int
+    yb: TD
+
+class TD(TDA, TDB):
+    a: TDA
+    b: TDB
+
+td: TD
+reveal_type(td)  # N: Revealed type is "TypedDict('__main__.TD', {'xb': builtins.int, 'yb': ..., 'xa': builtins.int, 'ya': ..., 'a': TypedDict('__main__.TDA', {'xa': builtins.int, 'ya': ...}), 'b': TypedDict('__main__.TDB', {'xb': builtins.int, 'yb': ...})})"
```
Comment (JukkaL) [inline test-data/unit/check-recursive-types.test:663]: Test constructing recursive TypedDicts, e.g. `TD(...)` (with different levels of nesting) and `td: TD = {...}`.

Test calling some methods on recursive typeddicts.

Test indexed assignment to a recursive typeddict.

File: test-data/unit/fine-grained.test [hunk 2]
```diff
@@ -3504,6 +3528,35 @@ def foo(x: Point) -> int:
 ==
 b.py:3: error: Unsupported operand types for + ("int" and "str")
 
+[case testTypedDictUpdate3]
```
Comment (JukkaL) [inline test-data/unit/fine-grained.test:3531]: Test coarse-grained incremental checking with recursive TypedDicts (i.e. serialization).

**Review by JukkaL — COMMENTED** [review]
Not a full review -- left some ideas about additional tests.

**Comment by ilevkivskyi** [comment]
Well it looks like merging didn't go perfect. Anyway the diff is correct, it is just the commit history shown includes commits from the previous PR. I added a bunch of tests as requested plus also some basic subtyping and joins.

**Review by JukkaL — APPROVED** [review]
LGTM, thanks for the PR! Recursive TypedDicts are a nice feature.

**Pull request merged**